File: core/entity/worker.js
"use strict";

var obj = function(map, config) {
    this.map = map;
    this.type = 'worker';
    this.pos = {
        x: config.x || 0,
        y: config.y || 0
    };
    this.health = 10;
    this.pulse = 0;
    this.nexus = config.nexus;
    this.side = config.side;
};
obj.prototype = require(appRoot + '/core/lib/extends.js')(require(appRoot + '/core/entity.js'), {
    think: function() {
        this.createPulse('nexus_' + this.side);

        this.nexus.health += 1;

        var ws = (this.nexus.food / this.nexus.water), cord = [[-1, 0], [1, 0], [0, -1], [0, 1]];

        var score = 0, next = null, type = '';
        for (var v in cord) {
            var c = cord[v], g = this.map.grid[(this.pos.x + c[0]) + '.' + (this.pos.y + c[1])];
            if (g && g.size > 0) {
                g.size -= 1;
                this.nexus[g.type] += 1;
                return;
            }

            var f = this.heat('food' + this.side, this.pos.x + c[0], this.pos.y + c[1]);
            if (f > score) {
                score = f;
                type = 'food' + this.side;
                next = [this.pos.x + c[0], this.pos.y + c[1]];
            }

            var w = this.heat('water' + this.side, this.pos.x + c[0], this.pos.y + c[1]) * ws;
            if (w > score) {
                score = w;
                type = 'water' + this.side;
                next = [this.pos.x + c[0], this.pos.y + c[1]];
            }
        }

        if (next) {
            this.setPos(next[0], next[1]);
            this.map.heat[type].set(this.pos.x, this.pos.y, 0);
        } else {
            var c = cord[Math.floor(Math.random() * cord.length)];
            this.setPos(this.pos.x + c[0], this.pos.y + c[1]);
        }
    }
});

module.exports = obj;

File: java/Worker.java

public class Worker extends Entity implements IEntity {
	private Map map;
	private int age;
	public Nexus nexus;
	
	protected Worker(Map m, Nexus n, float x, float y) {
		this.map = m;
		this.nexus = n;
		this.age = 0;
		this.x = x;
		this.y = y;
		log.write("log/game", "spawn worker " + n.side);
	}

	public Boolean think() {
		nexus.health = Math.min(100, nexus.health + 1);

		if (age > 500) {
			for(Object i : map.entity) {
				if(i instanceof Worker) {
					Worker w = (Worker)i;
					if (w == this) {
						map.entity.remove(i);
						return (false);
					}
				}
			}
		}
		this.age += 1;

		for(Object i : map.entity) {
			if(i instanceof Food) {
				Food f = (Food)i;
				if (f.stock > 0) {
					if (f.x == Math.round(x) && f.y == Math.round(y)) {
						f.stock -= 1;
						nexus.food += 1;
					} else {
						nextPos(1, f.x, f.y);
					}
					return (true);
				}
			}
			
			if(i instanceof Water) {
				Water w = (Water)i;
				if (w.stock > 0) {
					if (w.x == Math.round(x) && w.y == Math.round(y)) {
						w.stock -= 1;
						nexus.water += 1;
					} else {
						nextPos(1, w.x, w.y);
					}
					return (true);
				}
			}
		}

		nextPos(1, x + (Seed.random(20) - 10), y + (Seed.random(20) - 10));
		return (true);
	}
}
